### PR TITLE
TA as ELF + signature

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Changes format of TA to use the ELF format instead. This patch also adds
signature checking of the TA. The format of the TA is:
```
<Signed header>
<ELF header>
<Program header> (part of ELF spec, pointing out segments to load)
<Sections>
```

A struct ta_head is placed in the first section of the first segment
to carry flags and other properties of the TA.

elf32.h, elf64.h and elf_common.h are imported from FreeBSD.

In addition to the R_ARM_RELATIVE relocation type, adds support for
R_ARM_ABS32 relocations. Since a symbol table is needed to process this
relocation type a separate program header is added in the TA ELF
containing the sections .dynamic, .dynsym, .dynstr and .hash.  These
sections are only needed during relocation and could be released once
the TA is relocated.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU, FVP)

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Seems to have some checkpatch problems, please wait until fixed.

**Comment by jenswi-linaro** [comment]
All checkpatch warnings and errors that are relevant are fixed now as far as I can tell.
Please review.

**Comment by jenswi-linaro** [comment]
Depends on https://github.com/OP-TEE/optee_test/pull/22 for tests to compile.

**Comment by paswan-ms** [comment]
Re: the prior thread on the TA size increase. Whilst it's true that memory may be temporary in OpTEE, if I can figure out how purge the debug sections I think I'll need to do so because the TA binaries are packaged into our UEFI firmware - it's causing the size of the UEFI image to double and requires UEFI to need more memory to be able load the TA to pass to OpTEE when the session is opened.
Thanks,
Paul.

**Comment by jenswi-linaro** [comment]
Prior to signing the TA (in this PR) a new ELF is created from the original ELF with `objcopy --strip-unneeded`.  With this the new signed TAs disk usage grow by ~35% compared to the original "signed" TAs (based on values from the test TAs we have in optee_test.git).

**Comment by paswan-ms** [comment]
OK, thanks, I'll merge that in and try it.

**Comment by paswan-ms** [comment]
That worked - the image size is now back to normal. Our client is seeing a Free TA RPC call being made after the load but there isn't any debug trace showing that the load is failing - all the debug trace looks healthy. The memref in the Free TA RPC call has a valid buf_ptr but the size is zero, triggering an assertion in our client. I suspect it's overactive and it's OK for the size field to be zero in this call.
Thanks,
Paul.

File: core/arch/arm/kernel/elf_load.c [hunk 1]
```diff
@@ -0,0 +1,371 @@
+/*
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/elf_load.c]: FYI, I have not reviewed this file

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -236,373 +242,224 @@ static struct tee_ta_ctx *tee_ta_context_find(const TEE_UUID *uuid)
...
+	digest = malloc(shdr->hash_size);
+	if (!digest) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
 	}
-	memcpy(dst, nmem_ta, size);
 
-	cache_maintenance_l1(DCACHE_AREA_CLEAN, dst, size);
-	cache_maintenance_l1(ICACHE_AREA_INVALIDATE, dst, size);
+	res = crypto_ops.hash.final(hash_ctx, hash_algo, digest,
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: Can't we have the `final` and `memcmp` right after the `crypto_ops.hash.update`? In that case, can't we have it part of `tee_ta_load_check_shdr`?

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 3]
```diff
@@ -236,373 +242,224 @@ static struct tee_ta_ctx *tee_ta_context_find(const TEE_UUID *uuid)
...
+	/* man_flags: mandatory flags */
+	uint32_t man_flags = TA_FLAG_USER_MODE | TA_FLAG_EXEC_DDR;
+	/* opt_flags: optional flags */
+	uint32_t opt_flags = man_flags | TA_FLAG_SINGLE_INSTANCE |
+	    TA_FLAG_MULTI_SESSION | TA_FLAG_UNSAFE_NW_PARAMS;
 	struct tee_ta_ctx *ctx = NULL;
-	const uint8_t *nmem_ta;
-	kta_signed_header_t *sec_signed_ta = NULL;
-	ta_head_t *sec_head = NULL;
+	struct shdr *sec_signed_ta = NULL;
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: I would rename `sec_signed_ta` in `sec_signed_header_ta` as it will contain only the "sign/hash" part of the TA.

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 4]
```diff
@@ -626,57 +483,71 @@ static TEE_Result tee_ta_load(const kta_signed_header_t *signed_ta,
...
+
+	ctx->flags = ta_head->flags;
+
+	/* Check if multi instance && single session config  */
+	if (((ctx->flags & TA_FLAG_SINGLE_INSTANCE) == 0) &&
+	    ((ctx->flags & TA_FLAG_MULTI_SESSION) == 0)) {
+		/*
+		 * assume MultiInstance/SingleSession,
+		 * same as MultiInstance/MultiSession
+		 */
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: Why? I would say `MultiInstance/SingleSession` can be replaced by `SingleInstance/SingleSession`, which is more constrained.

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 5]
```diff
@@ -792,15 +657,15 @@ static TEE_Result tee_user_ta_enter(TEE_ErrorOrigin *err,
 	if (res != TEE_SUCCESS)
 		goto cleanup_return;
 
-	start_uaddr = ctx->load_addr + ta_func_head[func].start;
+	/*start_uaddr = ctx->load_addr + ta_func_head[func].start; */
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: Can be removed

File: core/tee/tee_svc.c [hunk 6]
```diff
@@ -313,17 +312,6 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
 					  param_types);
 	}
 
-	if ((called_sess != NULL) &&
```
Comment (pascal-brand-st-dev) [inline core/tee/tee_svc.c]: Is this change on purpose in this Pull-Request? It does not look related to ELF / signature changes

File: core/tee/tee_svc.c [hunk 7]
```diff
@@ -443,16 +431,13 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
...
 		struct tee_ta_param *param,
 		tee_paddr_t tmp_buf_pa[TEE_NUM_PARAMS],
 		struct abi_user32_param *usr_param)
 {
 	size_t n;
 	TEE_Param callee_params[TEE_NUM_PARAMS];
-	bool have_private_mem_map = (called_sess == NULL) ||
-		(called_sess->ctx->static_ta != NULL) ||
-		((called_sess->ctx->flags & TA_FLAG_USER_MODE) != 0);
+	bool have_private_mem_map = true;
```
Comment (pascal-brand-st-dev) [inline core/tee/tee_svc.c]: `have_private_mem_map = false` for kernel TA

File: lib/libutee/include/user_ta_header.h [hunk 8]
```diff
@@ -30,73 +30,30 @@
 
 #include <tee_api_types.h>
 
+/* XXX */
+#define TA_FLAG_USER_MODE 0
```
Comment (pascal-brand-st-dev) [inline lib/libutee/include/user_ta_header.h]: No more `TA_FLAG_USER_MODE`?

File: mk/compile.mk [hunk 9]
```diff
@@ -141,4 +141,8 @@ endef
 $(foreach f, $(srcs), $(eval $(call \
 	process_srcs,$(f),$(out-dir)/$(base-prefix)$$(basename $f).o)))
 
+# Handle generated source files, that is, files that are compiled from out-dir
+$(foreach f, $(gen-srcs), $(eval $(call \
```
Comment (pascal-brand-st-dev) [inline mk/compile.mk]: Should not we do the same for  `version.o `? Could be part of another Pull-Request not to mix feature changes (this PR is already big).

File: scripts/sign.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
```
Comment (pascal-brand-st-dev) [inline scripts/sign.py]: Would be good to have a .md file describing what is signed and how

File: ta/arch/arm/link.mk [hunk 11]
```diff
@@ -4,13 +4,14 @@ link-script = $(TA_DEV_KIT_DIR)/src/user_ta_elf_arm.lds
 link-script-pp = $(link-out-dir)/ta.lds
 
 FIX_TA_BINARY = $(TA_DEV_KIT_DIR)/scripts/fix_ta_binary
-
+SIGN = $(TA_DEV_KIT_DIR)/scripts/sign.py
```
Comment (pascal-brand-st-dev) [inline ta/arch/arm/link.mk]: do we still need fix_ta_binary?

File: ta/arch/arm/link.mk [hunk 12]
```diff
@@ -4,13 +4,14 @@ link-script = $(TA_DEV_KIT_DIR)/src/user_ta_elf_arm.lds
...
+TA_KEY = $(TA_DEV_KIT_DIR)/keys/default_ta.pem
```
Comment (pascal-brand-st-dev) [inline ta/arch/arm/link.mk]: * `?=` instead of `=`  ?
* rename in `TA_SIGN_KEY` as named in `mk/config.in` ?

File: ta/arch/arm/user_ta_header.c [hunk 13]
```diff
@@ -58,42 +58,19 @@ void ta_entry_invoke_command(uint32_t cmd_id, uint32_t param_types,
...
-	/* Section size information */
-	(uint32_t)&linker_RO_sections_size,
-	(uint32_t)&linker_RW_sections_size,
-	(uint32_t)&linker_res_funcs_ZI_sections_size,
-	(uint32_t)&linker_rel_dyn_GOT,
-	/* Hash type, filled in by sign-tool */
-	0,
-	/* TA trace level */
-	/* TA_TRACE_LEVEL_DEFAULT, */
+	TA_STACK_SIZE,
```
Comment (pascal-brand-st-dev) [inline ta/arch/arm/user_ta_header.c]: `TA_DATA_SIZE` has disappeared. Should not it be back as there is a GP property `gpd.ta.dataSize`?

File: ta/arch/arm/user_ta_header.c [hunk 14]
```diff
@@ -58,42 +58,19 @@ void ta_entry_invoke_command(uint32_t cmd_id, uint32_t param_types,
...
+	TA_FLAGS,
```
Comment (pascal-brand-st-dev) [inline ta/arch/arm/user_ta_header.c]: Can we still distinguish a UserTA from a Kernel TA?

**Comment by pascal-brand-st-dev** [comment]
2 general remarks:
* I understand that the way we know / use a TA is a user-TA or a kernel-TA can be simplified. But I would leave it for another patch as it is quite hard (for me) to know all the consequences inside this big PR
* Any possibility to have optional signature / check?

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: No, `elf_load_head()` and `elf_load_body()` does a few calls to `crypto_ops.hash.update()` via `copy_to()`.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: Good point,  I'll change it to `sec_shdr` if you don't mind.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 4 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: This is just copy paste from the old `tee_ta_load_user_ta()` function. If think we could remove this test all together as the multi-session property is ignored for for multi-instance TAs.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 5 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: Thanks

File: core/tee/tee_svc.c (hunk 6 above)
Comment (jenswi-linaro) [inline core/tee/tee_svc.c]: Except that I've cleaned out some other kernel TA leftovers in this patch. I can try to break out what's possible of that cleanup in a separate patch if you like.

File: core/tee/tee_svc.c (hunk 7 above)
Comment (jenswi-linaro) [inline core/tee/tee_svc.c]: Agree, but we don't have kernel TAs any longer. This variable should be removed either in this patch or in separate kernel TA cleanup patch whatever you prefer.

File: mk/compile.mk (hunk 9 above)
Comment (jenswi-linaro) [inline mk/compile.mk]: Yes let's wait with that.

File: lib/libutee/include/user_ta_header.h (hunk 8 above)
Comment (jenswi-linaro) [inline lib/libutee/include/user_ta_header.h]: All these TAs are user mode. We should probably remove the define completely as we're breaking backwards compatibility with this commit any way.

File: scripts/sign.py (hunk 10 above)
Comment (jenswi-linaro) [inline scripts/sign.py]: OK, I'll add a section under "11. Trusted Applications" in `optee_design.md`

File: ta/arch/arm/link.mk (hunk 11 above)
Comment (jenswi-linaro) [inline ta/arch/arm/link.mk]: No, I'll remove it.

File: ta/arch/arm/link.mk (hunk 12 above)
Comment (jenswi-linaro) [inline ta/arch/arm/link.mk]: OK, I'll fix

File: ta/arch/arm/user_ta_header.c (hunk 13 above)
Comment (jenswi-linaro) [inline ta/arch/arm/user_ta_header.c]: The heap is allocated in .bss instead now. The GP property should still be there.

File: ta/arch/arm/user_ta_header.c (hunk 14 above)
Comment (jenswi-linaro) [inline ta/arch/arm/user_ta_header.c]: No, we don't support kernel TAs. We haven't had support for kernel TAs from some time now.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 2 above)
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: There is no way to compute the hash in a single step? If possible, I would prefer as we would have a clean separation between the "check" part and the "load" part. But no problem if not possible or too messy.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 4 above)
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: Yes, let's remove it

File: ta/arch/arm/user_ta_header.c (hunk 13 above)
Comment (pascal-brand-st-dev) [inline ta/arch/arm/user_ta_header.c]: Ok

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: It is possible but that would mean that we need to first copy the entire ELF into secure DDR and verify the hash, after that we'll need to copy parts of the ELF into a new location in secure DDR. This isn't a problem for small TAs, but consider a large TA that would consume half of the available secure DDR. One would expect to be able to load such a TA, but that will not be possible here.

With the current code we copy parts of the ELF from shared memory piece by piece, only requiring the amount of secure DDR needed by the TA to execute.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 4 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: OK, I'll fix.

File: core/arch/arm/kernel/tee_ta_manager.c (hunk 2 above)
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/tee_ta_manager.c]: Ok, then let's forget about it.

**Comment by jenswi-linaro** [comment]
Addressed review comments

**Comment by jenswi-linaro** [comment]
Fixed a problem with locating the symbol table in the ELF.

**Comment by pascal-brand-st-dev** [comment]
Apart from the file `elf_load.c` not reviewed,

    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>
    Tested-by: Pascal Brand <pascal.brand@linaro.org> (STM platform)
    Tested-by: Pascal Brand <pascal.brand@linaro.org> (QEMU platform)

**Comment by jenswi-linaro** [comment]
`elf_load.c` need to be reviewed by someone too. Any volunteers to learn something about ELFs, @jforissier?

Link for ELF definition http://www.sco.com/developers/gabi/latest/contents.html
Link to ARM specific stuff for ELF http://infocenter.arm.com/help/topic/com.arm.doc.ihi0044e/index.html

**Comment by jforissier** [comment]
@jenswi-linaro yes I will check this, thanks for the links.

**Comment by jenswi-linaro** [comment]
Thanks @jforissier

File: core/arch/arm/kernel/elf_load.c [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+	    ehdr->e_ident[EI_DATA] != ELFDATA2LSB ||
+	    ehdr->e_ident[EI_OSABI] != ELFOSABI_NONE ||
+	    ehdr->e_type != ET_DYN || ehdr->e_machine != EM_ARM ||
+	    ehdr->e_phentsize != sizeof(Elf32_Phdr) ||
+	    ehdr->e_shentsize != sizeof(Elf32_Shdr))
+		return TEE_ERROR_BAD_FORMAT;
+
+	/*
+	 * Program headers are supposed to be arranged as:
+	 * LOAD [0] : .ta_head ...
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: Nit-picking: it may be better to use the full name of the ELF constants in the comments, that is: `PT_LOAD`, `PT_DYNAMIC`.

File: core/arch/arm/kernel/elf_load.c [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+	 */
+	if (ehdr->e_phnum < 1)
+		return TEE_ERROR_BAD_FORMAT;
+	res = alloc_and_copy_to(&p, state, ehdr->e_phoff,
+				ehdr->e_phnum * sizeof(Elf32_Phdr));
+	if (res != TEE_SUCCESS)
+		return res;
+	state->phdr = p;
+
+	if (state->phdr[0].p_type != PT_LOAD || state->phdr[0].p_vaddr != 0)
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: How do you know for sure that `p_vaddr` will be set to 0 by the linker? Is it implicitely done due to the linker script we are using for TAs?

File: core/arch/arm/kernel/elf_load.c [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+		return res;
+	state->ta_head = p;
+	state->ta_head_size = head_size;
+
+	*head = state->ta_head;
+	*vasize = state->vasize;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result el_process_rel(struct elf_load_state *state, size_t rel_sidx,
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: `elf_process_rel()` maybe?

File: core/arch/arm/kernel/elf_load.c [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+				ehdr->e_phnum * sizeof(Elf32_Phdr));
+	if (res != TEE_SUCCESS)
+		return res;
+	state->phdr = p;
+
+	if (state->phdr[0].p_type != PT_LOAD || state->phdr[0].p_vaddr != 0)
+		return TEE_ERROR_BAD_FORMAT;
+
+	/*
+	 * Calculate amount of required physical memory for TA. Find the
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: s/physical/virtual/ I guess?

File: core/arch/arm/kernel/elf_load.c [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+	 * max address used by a LOAD type. Note that last LOAD type
+	 * dictates the total ammount of needed memory. Eventual holes in
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: amount

File: core/arch/arm/kernel/elf_load.c [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+	 * as we already know from above that state->phdr[0].p_type == PT_LOAD
+	 */
+	n = ehdr->e_phnum - 1;
+	while (state->phdr[n].p_type != PT_LOAD)
+		n--;
+	state->vasize = state->phdr[n].p_vaddr + state->phdr[n].p_memsz;
+
+	/*
+	 * Read .ta_head from first segment, make sure the segment is large
+	 * enough. We're only interesting in seeing that the
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: interested

File: core/arch/arm/kernel/elf_load.c [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+	}
+
+	/*
+	 * We have now loaded all segments into TA memory, now we need to
+	 * process relocation information. To find relocation information
+	 * we need to locate the section headers. The section headers are
+	 * located somewhere between the last segment and the end of the
+	 * ELF.
+	 */
+	if (ehdr->e_shoff < phdr[1].p_filesz)
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: What if the program header table has only 1 entry (`ehdr->e_phnum == 1)`? `phdr[1]` would not exist. Anyway, is this test really needed?

File: core/arch/arm/kernel/elf_load.c [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+	res = advance_to(state, state->nwdata_len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* Find section containing symbol table */
+	n = 0;
+	while (true) {
+		if (n >= ehdr->e_shnum)
+			return TEE_ERROR_BAD_FORMAT;
+		if (state->shdr[n].sh_type == SHT_DYNSYM)
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: Or `SHT_SYMTAB`?

File: core/arch/arm/kernel/elf_load.c [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+	state->ehdr = ehdr;
+
+	if (!IS_ELF(*ehdr) ||
+	    ehdr->e_ident[EI_VERSION] != EV_CURRENT ||
+	    ehdr->e_ident[EI_CLASS] != ELFCLASS32 ||
+	    ehdr->e_ident[EI_DATA] != ELFDATA2LSB ||
+	    ehdr->e_ident[EI_OSABI] != ELFOSABI_NONE ||
+	    ehdr->e_type != ET_DYN || ehdr->e_machine != EM_ARM ||
+	    ehdr->e_phentsize != sizeof(Elf32_Phdr) ||
+	    ehdr->e_shentsize != sizeof(Elf32_Shdr))
```
Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: You might want to check that `(ehdr->e_flags & EF_ARM_ABI_FLOAT_HARD == 0)`, too.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: Will fix.

File: core/arch/arm/kernel/elf_load.c (hunk 15 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: Will fix.

File: core/arch/arm/kernel/elf_load.c (hunk 16 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: Yes, we're linking against address virtual address 0. If the TA would be linked against some other address the allocation of memory etc would be wrong as there would be a big hole at the beginning. It would be possible to handle that but the code would be slightly more complicate without gaining much.

I'll add a comment.

File: core/arch/arm/kernel/elf_load.c (hunk 18 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: You're right.

File: core/arch/arm/kernel/elf_load.c (hunk 21 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: You're right we can skip that (faulty) test.

I'll update the code to support absent section table (no relocation), it doesn't make things much more complicated even if it's very unlikely that we'll need it.

File: core/arch/arm/kernel/elf_load.c (hunk 22 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: I'll add code to fall back to `SHT_SYMTAB` if `SHT_DYNSYM` can't be found.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: Hmm, it wasn't that easy.

Comment (jforissier) [inline core/arch/arm/kernel/elf_load.c]: OK maybe it's not worth the trouble then. In which case would we have `SHT_SYMTAB` rather than `SHT_DYNSYM`?

**Comment by jenswi-linaro** [comment]
Addressed review comments. The way the symbol table is looked has changed too.

File: ta/arch/arm/link.mk [hunk 24]
```diff
@@ -42,7 +40,10 @@ $(link-out-dir)/$(binary).dmp: $(link-out-dir)/$(binary).elf
...
 	$(q)$(OBJDUMP$(sm)) -l -x -d $< > $@
 
-$(link-out-dir)/$(binary).bin: $(link-out-dir)/$(binary).elf
+$(link-out-dir)/$(binary).stripped.elf: $(link-out-dir)/$(binary).elf
 	@$(cmd-echo-silent) '  OBJCOPY $@'
-	$(q)$(OBJCOPY$(sm)) -O binary $< $@
-	$(q)$(FIX_TA_BINARY) $< $@
+	$(q)$(OBJCOPY$(sm)) --strip-unneeded $< $@
+
+$(link-out-dir)/$(binary).ta: $(link-out-dir)/$(binary).stripped.elf
```
Comment (jforissier) [inline ta/arch/arm/link.mk]: Missing dependency on `$(TA_SIGN_KEY)`

**Comment by jforissier** [comment]
For `elf_load.c`:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

Also, maybe you could add a few words somewhere (commit message? optee_design.md?) to explain where the default private key is located (`keys/default_ta.pem`) and how one may generate his own (`openssl genrsa -out keys/default_ta.pem`). Well it's up to you -- it wasn't too hard to figure out.

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tags applied.
Please note that this PR depends on https://github.com/OP-TEE/optee_test/pull/22 and vice versa.

**Comment by jenswi-linaro** [comment]
Are we done here? Should rebase again?

**Comment by pascal-brand-st** [comment]
Merged manually
